### DISCUSSION TITLE
[Symfony 5.3] Profiler updated/replaced after Ajax request

### DISCUSSION THREAD

**Discussion by meiyasan** [symfony/symfony#42439] [Q&A]
Hello guys,

I am trying to follow the tutorial : https://symfony.com/doc/current/profiler.html#updating-the-web-debug-toolbar-after-ajax-requests. I am actually expecting from this tutorial that the toolbar will be refreshing my profiler information (dump messages, and database accesses), after I sent my Ajax request. I have implemented a JS page transitioning using AJAX, so it would be very helpful to see the toolbar updated..

However I get hard time to make the subscriber working (cf. code below).
There are a few errors, that I tried to tackle :
- The method "$this->getKernel()" is not accessible. (this I could fix by injecting kernel in my subscriber)
- Then "$response->headers" is returning a notice message: `Notice: Trying to get property 'headers' of non-object` (There I just added an additional check verifying "$response" is not NULL)

Following this it seems my toolbar is not displayed anymore.
I assume I am missing some cases for the subscriber to not be set. (for instance in "/_wdt", "/_profile" routes)

Can someone help me with that? 
Thank you so much for your help :-)

Best regards,
Marco

```php
use Symfony\Component\HttpKernel\Event\ResponseEvent;

// ...

public function onKernelResponse(ResponseEvent $event)
{
    if (!$this->getKernel()->isDebug()) {
        return;
    }

    $request = $event->getRequest();
    if (!$request->isXmlHttpRequest()) {
        return;
    }

    $response = $event->getResponse();
    $response->headers->set('Symfony-Debug-Toolbar-Replace', 1);
}
```

**Comment by meiyasan** [comment]
Can someone confirm this header option is doing what I am expecting ?

**Reply by faizanakram99** [reply]
I personally don't care much about toolbar as profiler can be accessed by going to `/_profiler` route and there you can see all requests (including ajax requests). You can view any request in detail by clicking on it.

![image](https://user-images.githubusercontent.com/10880992/128818631-e3dfcf04-95d0-4aaa-9fb8-570dc8fbd7ae.png)

As you can see in the image, it also lists ajax routes like http://qt.test/js/routing?callback=fos.Router.setData

**Reply by meiyasan** [reply]
Is that a way to say that "profiler replace" flag is buggy ? 🤔

I actually care a lot, because I am running Js transition pages, so currently I only see the Xhr updating but I would like to also print out dump message. The profiler summary bar is actually very powerful. Then for further information, of course, I would go on the main profiler page

**Reply by faizanakram99** [reply]
The docs seem a bit outdated, it's assumed that event subscriber is container aware which is no more the case since symfony 3.

Maybe a good idea is to create a minimal reproducer, if profiler replace flag doesn't work with it as intended, then please open an issue.

Also, instead of injecting kernel, you can inject kernel.debug parameter.

How are you making Ajax request ?
Is XmlHttpRequest header set ?
(It's not set for fetch requests for example). In that case you can set a custom header and filter responses based on that custom header instead of checking for XmlHttpRequests

**Reply by meiyasan** [reply]
I am used to use jQuery.ajax()

**Comment by meiyasan** [comment]
Here is the solution I was expected.. when receiving an XHR response
Perhaps, this might be updated in the documentation.. because it is crazy helpful for me..
```

            if(xhr) {
                var xdebugToken = xhr.getResponseHeader('X-Debug-Token');

                // If the Sfjs object exists
                if (typeof Sfjs !== "undefined") {

                    // Grab the toolbar element
                    var currentElement = $('.sf-toolbar')[0];

                    // Load the data of the given xdebug token into the current toolbar wrapper
                    Sfjs.load(currentElement.id, '/_wdt/'+ xdebugToken);
                }
            }
            
```

**Reply by meiyasan** [reply]
Perhaps I can say a few additional words:
- Using this method the dropdown menu is not poping properly
- I can't close the toolbar clicking on the close button

I am trying to figure out why.. any help is welcome :-)

**Reply by meiyasan** [reply]
It seems to come from the fact that the "sftwdt-[0-9]*" identifier is changing name due to new token, while the closing button is not refreshing it's identifier name and keeps previous token naming.

This actually looks like a bug to me.. but perhaps someone more familiar with Sfjs function knows better about how to refresh the closing button and the drop down menu not fitting screen ?
